xml/hwbp_registers.xml

```diff
@@ -683,33 +683,42 @@
     <register name="Instruction Count" short="icount" address="0x7a1">
         This register is accessible as \RcsrTdataOne when \FcsrTdataOneType is 3.
 
-        \begin{commentary} This trigger type is intended to be used as a single
-        step that's useful both for external debuggers and for software monitor
-        programs. For that case it is not necessary to support \FcsrIcountCount greater
-        than
-        1. The only two combinations of the mode bits that are useful in those
-        scenarios are \FcsrIcountU by itself, or \FcsrIcountM, \FcsrIcountS, and \FcsrIcountU all set.
-
-        If the hardware limits \FcsrIcountCount to 1, and changes mode bits instead of
-        decrementing \FcsrIcountCount, this register can be implemented with just 2
-        bits. One for \FcsrIcountU, and one for \FcsrIcountM and \FcsrIcountS tied together.  If
-        only the external debugger or only a software monitor needs to be
-        supported, a single bit is enough.
+        If \FcsrIcountCount is not 0, then every instruction completed or
+        trap taken from a mode where the trigger is enabled decrements
+        \FcsrIcountCount by 1, unless \FcsrIcountAction is 0, the current
+        mode is M-mode, and M-mode breakpoints are disabled via one of the
+        mechanisms described in Section~\ref{sec:mmtrigger}. When
```

Comment:
After this change introducing the pending bit, I wonder if this is even necessary anymore. I had suggested this behavior since otherwise the breakpoint exception would get lost when the count reached 0 and exceptions were blocked. But now with the pending bit, that exception would not get lost anymore.

Reply:
I don't quite follow you. I thought pending is orthogonal to the mechanisms that might disable triggers in M-mode. Can you explain a bit more what you mean?

Reply:
This change (to stop count decrementing when exceptions were blocked) was #555. I opened that issue because at the time, if count decremented to 0 and breakpoint exceptions were blocked, the breakpoint would get dropped on the floor and never taken. Stopping the decrementing when exceptions were blocked was a way to avoid dropping that breakpoint exception completely.

But now, with this new pending bit, if the count still decrements even when exceptions are blocked, the pending bit will turn on and stay on until the exception can be taken, so it will no longer get dropped on the floor.

So I think #555 is moot and should be undone.

Reply:
I agree. There is maybe some use to disabling counting while in M-mode, but it makes the language more cumbersome and presumably the implementation as well. For the single step case #555 is no longer needed, and I cannot really imagine why somebody would want to count M-mode instructions except the ones executed while M-mode triggers are enabled.